Introduce Response.body. (#118)

yutakahirano opened the pull request:
This change introduces `body` property of Response class. Note that we don't
introduce `Request.body` at this moment and hence some properties and algorithms
defined in the Body mixin are moved to Request and Response.

`bodyUsed` property is now based on `IsDisturbed` defined in the streams spec.
Request.bodyUsed emulates the behavior without using streams.

yutakahirano commented:
html diff: http://services.w3.org/htmldiff?doc1=https%3A%2F%2Ffetch.spec.whatwg.org%2F&doc2=https%3A%2F%2Fraw.githubusercontent.com%2Fyutakahirano%2Ffetch%2Fresponse-body%2FOverview.html#dom-response

annevk commented:
What is the reason for not introducing `Request.prototype.body`?

domenic commented on Overview.html:
```diff
@@ -4080,11 +4131,14 @@ <h3 id="response-class"><span class="secno">6.4 </span>Response class</h3>
+   <li><p>Set <var title="">r</var>'s
+   <a href="#concept-response-readable-stream" title="concept-Response-readable-stream">readable stream</a> to the result of
+   <a href="#concept-construct-fixed-readablestream" title="concept-construct-fixed-ReadableStream">constructing</a> a fixed
+   <a href="#concept-readablestream" title="concept-ReadableStream">ReadableStream</a> with an array consisting of a
```
The input to the constructing a readable stream algorithm isn't a JS array, but instead an abstract list type. (In ES it would be a List, but most other specs including fetch are a bit less formal.) I'd say either "with a Uint8Array..." or "with a list whose single element is a Uint8Array..."

yutakahirano replied:
Done.

domenic commented on Overview.src.html:
```diff
@@ -3965,9 +3962,63 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+<code>ReadableByteStream</code> has all methods and properties which
+<span title=concept-ReadableStream>ReadableStream</span> has.
+
+<p>A <code>Response</code> object also has an associated
```
A better phrasing would be "A Response object is said to be locked if its associated readable stream..." Similarly for disturbed.

Although maybe these concepts should apply to abstract "responses" and not to Response objects? I am not sure.

domenic replied:
Sorry, disturbed is a bit different. It is mostly fine as you defined it, except I think it should probably apply to interface types, not to individual objects. But I guess the spec as it is currently defines them on individual objects, so no need to change that in your patch.

I think locked would be clearer as I suggested though, unless I am missing something.

yutakahirano replied:
Fixed the definition of "locked".

domenic commented on Overview.html:
```diff
@@ -3415,119 +3415,84 @@ <h3 id="body-mixin"><span class="secno">6.2 </span>Body mixin</h3>
-<dfn id="concept-body-consume-body" title="concept-Body-consume-body">consume body</dfn> algorithm, which given a
-<var title="">type</var>, runs these steps:
+<dfn id="concept-body-package-data" title="concept-Body-package-data">package data</dfn> algorithm, which given a
+<var title="">bytes</var>, a <var title="">type</var> and a <var title="">MIME type</var>, runs these steps:
```
given bytes, not given a bytes

yutakahirano replied:
Done.

domenic commented on Overview.src.html:
```diff
@@ -3965,9 +3962,63 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+
+ <li><p>Let <var title>p</var> be a new promise.
+
+ <li><p>Let <var title>promise</var> be the result of
```
I think this section would be clearer if instead of p and promise, you just said "Let promise be the result of reading all bytes from stream with reader. Return the result of transforming promise by a fulfillment handler that ...". It translates more naturally to idiomatic promise code that way too.

yutakahirano replied:
Done.

domenic commented on Overview.src.html:
```diff
@@ -3965,9 +3962,63 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+
+<p class = "note no-backref">The type of the associated
+<span title=concept-Response-readable-stream>readable stream</span> is defined as
+<span title=concept-ReadableStream>ReadableStream</span>, but it is currently discussed. Depending
```
insert "being" between currently and discussed.

yutakahirano replied:
Done.

yutakahirano commented:
> What is the reason for not introducing Request.prototype.body?

Hmm, I made that decision ago, and at that time it looked difficult to reading body stream bytes when fetching. I reconsidered it and it might be possible by changing the fech() function as following.

from:

 - Fetch _request_

to:

 - Let _reader_ be the result of getting a reader from _request_.body. If that throws an exception, reject _p_ with that exception and abort these steps.
 - Let _promise_ be the result of reading all data from _request_.body with _reader_.
  - When _promise_ is fulfilled with _bytes_, set _request_'s body to _bytes_ and Fetch _request_.
  - When _promise_ is rejected with an error, reject _p_ with that error.

What do you think about it?

annevk commented:
That does not seem ideal since it doesn't stream.

yutakahirano commented:
> That does not seem ideal since it doesn't stream.

I agree, and addressing the problem requires much effort, because it needs a fetch algorithm modification. There is a [ticket](https://github.com/yutakahirano/fetch-with-streams/issues/41) for the problem and I marked it as "not included in the first release".

annevk commented:
Okay, fair enough I guess.

yutakahirano commented:
Thanks, then I will include Request.body in this change (I'm on vacation and will be back on Sep 28).

annevk commented:
Sorry, I think you misunderstood. I agree with you that it's better to wait we have actual streaming sorted.

yutakahirano commented:
Oops, sorry for my misunderstanding. Let me confirm, Is it OK to 

1. Land this change (with addressing reviewers' comments) and
1. Address https://github.com/yutakahirano/fetch-with-streams/issues/41 after that ?

annevk commented:
Yes.

yutakahirano commented:
Thank you!

annevk commented on Overview.src.html:
```diff
@@ -3355,119 +3355,84 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
 <dfn title=concept-Body-MIME-type>MIME type</dfn> (initially the empty byte sequence).
 
+<p>Objects implementing the <code>Body</code> mixin must define an associated
+<dfn title=concept-Body-disturbed>disturbed</dfn> predicate and a
```
What is a predicate? Can we turn this into a flag?

domenic replied:
A predicate is a function that returns a boolean. It seems like a better model than a flag in this case, since you want it to be re-computed every time it is consulted.

annevk replied:
Hmm mkay.

annevk commented on Overview.src.html:
```diff
@@ -3355,119 +3355,84 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-  </ol>
+  <dt><i title>JSON</i>
+  <dd>
+   <p>Let <var title>JSON</var> be the result of invoking the initial value of the
```
Shouldn't you return the result?

yutakahirano replied:
Thanks, done.

annevk commented on Overview.src.html:
```diff
@@ -3557,9 +3522,40 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+
+ <li><p>Let <var title>p</var> be a new promise.
+
+ <li><p>Run these substeps <span data-anolis-spec=html>in parallel</span>:
```
Newline after `<li>` since it has more than one child.

yutakahirano replied:
Done. Is the indentation correct?

annevk commented on Overview.src.html:
```diff
@@ -4167,40 +4235,90 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
    <span title=concept-Headers-guard>guard</span> is "<code title>immutable</code>".
+
+   <li><p>If <var title>response</var>'s status is a <span>null body status</span>, resolve
+   <var title>p</var> with <var title>res</var> and abort these steps.
```
I don't think this is needed here. This is already enforced within Fetch, no?

yutakahirano replied:
This is related to the below comment. If &lt;concept-response-body&gt; is a ReadableStream, we don't need this item. Otherwise, we need this early-exit to make `res` have null body.

annevk commented on Overview.src.html:
```diff
@@ -4167,40 +4235,90 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <var title>stream</var>.
 
-  <p>To <span>process response body</span> for <var title>response</var>, do nothing.
+   <li><p>Clear out <var title>response</var>'s <span title=concept-response-body>body</span>.
```
I don't like this. We should define this more naturally. concept-response-body should be a stream that simply pipes into this one or some such.

yutakahirano replied:
Hmm, that is [another TODO] (https://github.com/yutakahirano/fetch-with-streams/issues/40). In my plan I would do the refactoring after landing this change. Replacing &lt;concept-response-body&gt; will impact other specs in non-trivial ways (https://xhr.spec.whatwg.org/#dom-xmlhttprequest-send for example), do you want to do it here?

annevk commented:
Also, there's no need to use `<var title>` anymore, just use `<var>` for new text (and when changing existing text).

yutakahirano commented:
> Also, there's no need to use `<var title>` anymore, just use `<var>` for new text (and when changing existing text).

Fixed added / modified ones. How about &lt;code title&gt; and &lt;code title=""&gt;?

annevk commented:
`<code title>` is still needed as `<code>` still has linking semantics.

annevk commented on Overview.src.html:
```diff
@@ -3557,9 +3520,41 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+ <li>
+  <p>Run these substeps <span data-anolis-spec=html>in parallel</span>:
+  <ol>
+   <li><p>Let <var>bytes</var> be <span title=concept-Request-request>request</span>'s
```
Shouldn't this do a bit more? E.g., read/consume the body somehow since it's a stream?

yutakahirano replied:
Maybe I don't understand what "byte stream" is. What operations does it have?

"Let _bytes_ be the result of consuming request's body as a byte sequence" makes sense?

annevk replied:
"Let bytes be the result of reading from stream until it returns end-of-stream." is what I use now, mostly based on https://encoding.spec.whatwg.org/#concept-stream and waiting for the Streams standard to become ready.

yutakahirano replied:
Fixed.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3975,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-<span title=concept-Response-response>response</span>'s
-<span title=concept-response-body>body</span>.
+<p>A <code>Response</code> object also has an associated
+<dfn title=concept-Response-readable-stream>readable stream</dfn> of type
```
I think I would prefer "(null or a ReadableStream object)" as done elsewhere. Or perhaps "which is null or a ReadableStream object". And then the initial value separately.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3975,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+Depending on the discussion, the type might be changed to <code>ReadableByteStream</code>. Hence the
+type name is not strictly specified at this moment, though the behavior is specified because
+<code>ReadableByteStream</code> has all methods and properties which
+<span title=concept-ReadableStream>ReadableStream</span> has.
```
I think it would be clearer to have something like `<p class=XXX><a href={linktoissue}>This might become a <code>ReadableByteStream</code></a>.`

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3975,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+ <li><p>Return the result of transforming <var>promise</var> by a fulfillment handler that
+ returns the result of the <span title=concept-body-package-data>package data</span> algorithm
+ with its first argument, <var>type</var> and <span title=concept-body-mime-type>MIME type</span>.
+</ol>
```
I don't really understand what transforming a promise means. @domenic?

domenic replied:
http://www.w3.org/2001/tag/doc/promises-guide#shorthand-reacting

annevk replied:
Hmm, we really need IDL to define this terminology.

annevk replied:
Can you add `<!-- XXX IDL really needs to define "transforming". For now it is defined in http://www.w3.org/2001/tag/doc/promises-guide -->` (wrapped if it exceeds the line length, aligned as per other comments in the document) please?

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4233,92 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-   "<code title>error</code>", reject <var title>p</var> with a
-   <code title>TypeError</code>.
+   <li><p>If <var>response</var>'s <span title=concept-response-type>type</span> is
+   "<code title>error</code>", reject <var>p</var> with a <code title>TypeError</code> and abort
```
abort -> terminate.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4233,92 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <var>response</var> and a new <code>Headers</code> object whose
    <span title=concept-Headers-guard>guard</span> is "<code title>immutable</code>".
+
+   <li><p>If <var>response</var>'s status is a <span>null body status</span>, resolve <var>p</var>
```
xref status. Having said that, I don't think we need this step at all. Why do you think we need it?

yutakahirano replied:
Without the early-exit, `res` has an empty associated readable stream instead of null. I stopped using null body status here and check `response`'s body's nullity instead.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4233,92 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <span title=concept-fetch-terminate>terminates</span> the ongoing fetch with reason
+   <i title>end-user abort</i>.
+
+   <li><p>Let <var>strategy</var> be an object. The user agent may choose any object.
```
Newline after `<li>` since it has two children.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4233,92 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <p class="note no-backref"><var>strategy</var> is used to control the queuing strategy of
+   <var>stream</var> constructed below.
+
+   <li><p>Set <var>stream</var> to the result of
```
New line after `<li>` since it has two children. Same below.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4233,92 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <var>stream</var>.
 
-  <p>To <span>process response body</span> for <var title>response</var>, do nothing.
+   <li><p>Clear out <var>response</var>'s <span title=concept-response-body>body</span>.
```
I don't understand this. "body" is itself a stream of sorts. Why would it be cleared?

yutakahirano replied:
I thought it might be unclear if reading is an idempotent operation or a destructive operation. If it's clear that reading from a byte stream is destructive, we can remove the sentence. Done.

annevk commented on Overview.src.html:
```diff
@@ -3355,119 +3355,82 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-       from the parsing operation, to
-       <span data-anolis-spec=xhr title=concept-FormData-entry>entries</span>.
-      </ol>
+ <li><p>Switch on <var>type</var>:
```
This seems to be the only step in this list? I think would be better to avoid the `<ol>` then and say something along the lines of "switch on type and run the associated steps" in the previous paragraph.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3355,119 +3355,82 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+  <dt><i title>FormData</i>
+  <dd>
+   <p>If <var>MIME type</var> (ignoring parameters) is `<code>multipart/form-data</code>`,
+   run these subsubsteps:
```
These are now just substeps.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3557,9 +3520,43 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <li><p>Resolve <var>p</var> with the result of running the
+   <span title=concept-body-package-data>package data</span> algorithm with <var>bytes</var>,
+   <var>type</var> and <span title=concept-Body-MIME-type>MIME type</span>. If that threw an
+  exception, reject <var>p</var> with that exception.
```
Indentation for "exception" is wrong.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3977,51 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-<span title=concept-response-body>body</span>.
+<p>A <code>Response</code> object also has an associated
+<dfn title=concept-Response-readable-stream>readable stream</dfn> which is null or a
+<span title=concept-ReadableStream>ReadableStream</span>. Its initial value is null.
```
ReadableStream is an object, right? Should we not use `<code>` for it and follow it with the word object? E.g., `<code title=...>ReadableStream</code> object`?

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3977,51 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+
+<p class="note no-backref">The type might be changed, but the behavior is specified because
+<code>ReadableByteStream</code> has all methods and properties which
+<span title=concept-ReadableStream>ReadableStream</span> has.
```
This can be folded into the XXX.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4034,11 +4074,14 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+   <span title=concept-Response-readable-stream>readable stream</span> to the result of
+   <span title=concept-construct-fixed-ReadableStream>constructing</span> a fixed
+   <span title=concept-ReadableStream>ReadableStream</span> with a list whose single element is a
+   <code>Uint8Array</code> whose contents are <var>bytes</var>. Rethrow any exceptions.
```
We should probably also say object after `Uint8Array`. Also, instead of "whose contents are" I think "whose buffer is" is more accurate. Do we want to spell out that the buffer is a new `ArrayBuffer`? I think we might. It seems you are already handling that throwing.

annevk replied:
Also, do we really want to require this consuming to happen during the constructor? It seems to throw away some of the benefits of a stream being asynchronous.

yutakahirano replied:
Hm, I added "constructing a ReadableStream with a byte stream" operation and use it here.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4234,93 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
    <span title=concept-Headers-guard>guard</span> is "<code title>immutable</code>".
+
+   <li><p>If <var>response</var>'s <span title=concept-response-body>body</span> is null, resolve
+   <var>p</var> with <var>res</var> and terminate these steps.
```
Then p never gets returned.

yutakahirano replied:
Here is in the "process response for response" handler, so I thought the termination was scoped in it (as "return" in the "process request body" handler in the XHR spec).

annevk replied:
That could work. But then use the "substep" (or subsubstep depending on the level of nesting) wording.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4234,93 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+    <span title=concept-construct-ReadableStream>constructing</span> a
+    <span title=concept-ReadableStream>ReadableStream</span> with <var>strategy</var>,
+    <var>pull</var> and <var>cancel</var>. If that threw an exception, run the following
+    substeps:
```
Newline.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -1073,7 +1073,7 @@ <h4 id=terminology-headers>Headers</h4>
 
 <p>To <dfn title=concept-enqueue-ReadableStream>enqueue</dfn> a byte sequence
-(<var title>chunk</var>) into a <span title=concept-ReadableStream>ReadableStream</span>
+(<var title>chunk</var>) into a <code title=concept-ReadableStream>ReadableStream</code>
```
`ReadableStream` object* (throughout)

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -1131,8 +1131,39 @@ <h4 id=terminology-headers>Headers</h4>
 
+<p>To <dfn title=concept-construct-ReadableStream-with-byte-stream>construct</dfn> a
+<code title=concept-ReadableStream>ReadableStream</code> with given a byte stream
+<var>bstream</var>, run these steps:
```
`<var>byteStream</var>`*

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -1131,8 +1131,39 @@ <h4 id=terminology-headers>Headers</h4>
+
+    <li><p>If <span data-anolis-spec=encoding title=concept-stream-read>read</span> returns
+    <span data-anolis-spec=encoding>end-of-stream</span>,
+    <span title=concept-close-ReadableStream>close</span> <var>stream</var> defined below.
```
This reads rather weird. Perhaps introduce a flag before these substeps, set it here, and then after these substeps and after `<var>stream</var>` is allocated, use it to close it if the flag is set?

yutakahirano replied:
_pull_ is called multiple times and end-of-stream should be checked every time.

annevk replied:
Oh I see, but then `pull` is presumably passed some arguments? That would make this more clear.

yutakahirano replied:
In fact, byte stream lacks some signals required by ReadableStream such as "process response body" and "process response end-of-file". I'm afraid we cannot specify this operation perfectly. Is it possible to state something like

To __construct__ a ReadableStream object from a byte stream _byteStream_, return a ReadableStream object so that consuming the ReadableStream object will result in chunks each of which is a Uint8Array object wrapping an ArrayBuffer and concatenate them all will result in a byte sequence that is equivalent to the byte sequence which would be read from _byteStream_.

?

annevk replied:
I guess so. Is there a long term plan to improve this?

yutakahirano replied:
I will use ReadableStream instead of byte stream in request and response.

For extracting Strings, ArrayBuffers and FormData objects, they can be converted to byte sequences and thus easily enqueued to a ReadableStream object.

For extracting Blobs, it would be possible to define an operation returning a ReadableStream object consisting of the blob contents (At least, as [blob] (https://w3c.github.io/FileAPI/#blob) refers to a byte sequence, enqueuing it to the ReadableStream object is possible. Allowing the UA to split the sequence to chunks would be beneficial, though).

yutakahirano replied:
Done.

annevk replied:
Yeah only using `ReadableStream` makes sense. That or an equivalent abstract concept.

annevk commented on Overview.src.html:
```diff
@@ -3355,120 +3386,81 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
-<dfn title=concept-Body-consume-body>consume body</dfn> algorithm, which given a
-<var title>type</var>, runs these steps:
+<dfn title=concept-Body-package-data>package data</dfn> algorithm, which given
+<var>bytes</var>, a <var>type</var> and a <var>MIME type</var>, switches on <var>type</var> and runs
```
comma after `<var>type</var>`.

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4034,11 +4103,14 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
 
-   <li><p>Let <var title>stream</var> and <var title>Content-Type</var> be the result of
-   <span title=concept-BodyInit-extract>extracting</span> <var title>body</var>.
+   <li><p>Let <var>bstream</var> and <var>Content-Type</var> be the result of
```
byteStream

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -1106,20 +1106,20 @@ <h4 id=terminology-headers>Headers</h4>
 
 <p>To <dfn title=concept-construct-fixed-ReadableStream>construct</dfn> a fixed
-<span title=concept-ReadableStream>ReadableStream</span> with given <var title>chunks</var>,
+<code title=concept-ReadableStream>ReadableStream</code> with given <var title>chunks</var>,
```
object*

yutakahirano replied:
Done.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4248,96 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+    </ol>
+
+   <li>
+    <p>Otherwise, run the following substeps:
```
This "Otherwise" is a tad confusing since the previous step doesn't start with an "If". Since the previous step ends up aborting the substeps, do we need an otherwise and substeps here?

yutakahirano replied:
Terminating a fetch operation and terminating substeps are different things. We should state both in this case. Fixed.

annevk commented on Overview.src.html:
```diff
@@ -4178,44 +4248,96 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
 
-   <li><p>Otherwise, resolve <var title>p</var> with a new <code>Response</code> object
-   associated with <var title>response</var> and a new <code>Headers</code> object whose
+   <li><p>Let <var>res</var> be a new <code>Response</code> object associated with
```
Perhaps we should call this variable `responseObject` or some such? Or maybe not rename the internal response from `r` to `response`...

yutakahirano replied:
I don't like `r` because it can be read as either one of request, response, Request and Response. I also want to use consistent names in this function - If I use `request` for a request, I wanna use `response` for a response and vice versa.

`responseObject` sounds good. Fixed.

annevk commented on Overview.src.html:
```diff
@@ -3979,9 +3990,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
+<code title=concept-ReadableStream>ReadableStream</code> object. Its initial value is null.
+
+<p class="XXX">This might become a <code>ReadableByteStream</code> object, depending on the
+<a href="https://github.com/whatwg/streams/issues/379">the discussion</a>. The type might be
```
"the the discussion" is wrong. I think this entire thing can be clearer as:

`<a>This might become a ReadableByteStream object</a>. While the type might change, the behavior specified will be equivalent since the hypothetical ReadableByteStream object will have the same members as the ReadableStream object has today.`

yutakahirano replied:
Oops, sorry. Fixed.

annevk commented on Overview.src.html:
```diff
@@ -3355,120 +3370,81 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
- <li><p>Let <var title>p</var> be a new promise.
+<dl class=switch>
+ <dt><i title>ArrayBuffer</i>
+ <dd><p>Return an <code>ArrayBuffer</code> whose contents are <var>bytes</var>.
```
Do you want to fix #134 here as well? It's okay if you don't, but with your new setup it seems fairly easy to just add the rethrow sentence.

yutakahirano replied:
Are you OK with https://github.com/yutakahirano/fetch/commit/bf979b16cf394dc0b84ef2716dbce4b264800d8a?

annevk replied:
Yes.

annevk commented:
Apart from that this looks good. Do you want to squash the commits and rebase or should I?

yutakahirano commented:
Thank you very much for the review!
I squashed and rebased the commits.

yutakahirano commented:
Addressing #118

annevk commented:
Thank you for the great patch!
